### PR TITLE
Refactor test POST code into a single method that is called several times

### PR DESCRIPTION
**Pull request by gregcorbett**
This removes the repeated

`response = test_client.post(url, MESSAGE, content_type="text/plain", HTTP_EMPA_ID="Test Process",  SSL_CLIENT_S_DN=example_dn)`

and replace them with a function call e.g.:

`self._make_record_post(MESSAGE, 202, dn=example_dn)`

This will also make it easier to test edge cases of a successful POST in future

### CODE REVIEW

**Comment by tofu-rocketry** [comment]
There's meant to be an extra comma in the arguments for the example above, right?

**Comment by gregcorbett** [comment]
Yes,

`self._make_record_post(MESSAGE, 202 dn=example_dn)`

should be

`self._make_record_post(MESSAGE, 202, dn=example_dn)`

**Comment by tofu-rocketry** [comment]
Looks like the slight reduction in coverage is due to the reduction in total lines as you've been removing duplication.

File: api/tests/test_cloud_record.py [hunk 1]
```diff
@@ -170,6 +102,57 @@ def _saved_messages(self, message_path):
         """Return a list of messages under message_path."""
         return glob.glob(message_path)
 
+    def _make_record_post(self, message, expected_status,
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: I feel this should be called something like `_check_record_post` as the asserts are really the key bit.

File: api/tests/test_cloud_record.py [hunk 2]
```diff
@@ -170,6 +102,57 @@ def _saved_messages(self, message_path):
...
+                          dn='/C=XX/O=XX/OU=XX/L=XX/CN=allowed_host.test',
+                          empaid='Test Process'):
+        """
+        Helper method to make a POST request.
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: Needs more on how this method can be used and what the expected results are.

File: api/tests/test_cloud_record.py [hunk 3]
```diff
@@ -170,6 +102,57 @@ def _saved_messages(self, message_path):
...
+
+        By default,
+         - dn is /C=XX/O=XX/OU=XX/L=XX/CN=allowed_host.test
+         - empaid is 'Test Process'
+        """
+        with self.settings(QPATH=QPATH_TEST):
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: What's the purpose of this?

File: api/tests/test_cloud_record.py [hunk 4]
```diff
@@ -91,70 +69,24 @@ def test_cloud_record_post_403(self):
...
-            # get message content
-            # can unpack sequence because we have asserted length 1
-            [message] = messages
-            message_file = open(message)
-            message_content = message_file.read()
-            message_file.close()
-
-            # check saved message content
-            self.assertEqual(MESSAGE, message_content)
+        # Mock the functionality of the provider list
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: Why?

File: api/tests/test_cloud_record.py [hunk 5]
```diff
@@ -170,6 +102,57 @@ def _saved_messages(self, message_path):
...
+         - empaid is 'Test Process'
+        """
+        with self.settings(QPATH=QPATH_TEST):
+
+            # Make the POST request
+            test_client = Client()
+            url = reverse('CloudRecordView')
+
+            if dn is not None:
+                # Include SSL_CLIENT_S_DN header
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: Why?

File: api/tests/test_cloud_record.py [hunk 6]
```diff
@@ -170,6 +102,57 @@ def _saved_messages(self, message_path):
...
+                response = test_client.post(url,
+                                            message,
+                                            content_type="text/plain",
+                                            HTTP_EMPA_ID=empaid,
+                                            SSL_CLIENT_S_DN=dn)
+            else:
+                # Omit SSL_CLIENT_S_DN header
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: Why?

File: api/tests/test_cloud_record.py [hunk 7]
```diff
@@ -170,6 +102,57 @@ def _saved_messages(self, message_path):
...
+                                            content_type="text/plain",
+                                            HTTP_EMPA_ID=empaid)
+
+            # check the expected response code has been received
+            self.assertEqual(response.status_code, expected_status)
+
+            # If the message was expected to be successfull, check for
+            # equality between the sent message and the saved message
+            if expected_status == 202:
+                # get save messages under QPATH_TEST
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: "saved messages"

File: api/tests/test_cloud_record.py (hunk 1 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: done

File: api/tests/test_cloud_record.py (hunk 2 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: done

File: api/tests/test_cloud_record.py (hunk 3 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: done

File: api/tests/test_cloud_record.py (hunk 5 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: added " Include the dn in SSL_CLIENT_S_DN so the request is able to be authenticated"

File: api/tests/test_cloud_record.py (hunk 6 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: added "Omit the SSL_CLIENT_S_DN header so the request is made without any certificate, and not authenticated"

File: api/tests/test_cloud_record.py (hunk 7 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: done

File: api/tests/test_cloud_record.py (hunk 4 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: added explanation

**Comment by gregcorbett** [comment]
I've updated the PR

File: api/tests/test_cloud_record.py [hunk 8]
```diff
@@ -91,70 +69,25 @@ def test_cloud_record_post_403(self):
...
-            [message] = messages
-            message_file = open(message)
-            message_content = message_file.read()
-            message_file.close()
-
-            # check saved message content
-            self.assertEqual(MESSAGE, message_content)
+        # Mock the functionality of the provider list,
+        # used in the underlying POST handling method so that the
+        # POST request is authorized, as allowed_host.test (the
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: Do you mean "as" or "so that"?

File: api/tests/test_cloud_record.py [hunk 9]
```diff
@@ -170,6 +103,73 @@ def _saved_messages(self, message_path):
...
+        # This avoids the test writing to the QPATH
+        # set in apel_rest/settings.py
+        with self.settings(QPATH=QPATH_TEST):
+
+            # Make the POST request
+            test_client = Client()
+            url = reverse('CloudRecordView')
+
+            if dn is not None:
+                # Include the in dn SSL_CLIENT_S_DN so the
```
Comment (tofu-rocketry) [inline api/tests/test_cloud_record.py]: "the DN in..."

File: api/tests/test_cloud_record.py (hunk 8 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: I think I mean "because", I've reworded it to hopefully make it clearer

File: api/tests/test_cloud_record.py (hunk 9 above)
Comment (gregcorbett) [inline api/tests/test_cloud_record.py]: changed

**Comment by gregcorbett** [comment]
I think it's supposed to be `subject`, as in certificate subject. There is also `SSL_CLIENT_I_DN` for the client certificate issuer, among other `SSL_CLIENT_*` variables.

Changes made
